Eng/pr cas support next

cachemeifyoucan opened the pull request.

akyrtzi committed:
[clang-cache-build-session] Replace the `clang-cache-build-session` script with a clang tool

This enables building more functionality on top of `clang-cache-build-session`.

Note that a minor downside of `clang-cache-build-session` being a tool, instead of a shell script, is that it
requires that shell scripts are annotated with a hashbang (`#!`) at the top, in order to be able to run them.

akyrtzi committed:
[clang-cache-build-session] Provide a way to infer the prefix mapping from a CMake build folder

Introduces `-prefix-map-cmake` flag which instructs `clang-cache-build-session` that it is running within a CMake
build directory that was enabled for [CMake's File API](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html).
`clang-cache-build-session` will then infer the CMake project's source paths and
setup `CLANG_CACHE_PREFIX_MAPS` environment variable for `clang-cache` to create prefix mapping
that results in build and source directory independent compilations.

In order to enable CMake's File API run

```
% mkdir -p .cmake/api/v1/query/codemodel-v2
```

inside the build directory before invoking the `cmake` configure command; this
only needs to happen once when the build directory is first created.

akyrtzi committed:
[cache-build-session] Move the `clang-cache-build-session` to llvm as `cache-build-session` tool

This tool is not dependent on clang and could also be useful for `swift` or other cctools during a CMake build.
As part of moving it to llvm, also change the environment variable names to use `LLVM_CACHE_` prefix.

cachemeifyoucan committed:
[CAS] Add CAS support

Add CAS support to llvm.

cachemeifyoucan committed:
[CAS] Add CAS support in clang

cachemeifyoucan committed:
[CAS] Add CMake support for build with CAS options

cachemeifyoucan commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -6530,7 +6530,10 @@ void CodeGenFunction::EmitDeclMetadata() {
+
+  // FIXME: Choose IncludeRevision based on command-line argument. Including
+  // the Git revision is bad for reproducible builds / caching.
+  std::string Version = getClangFullVersion(/*IncludeRevision=*/false);
```
Do we need to upstream this?

akyrtzi replied:
It's not clear to me we need this, people (including us) should be living on releases, not a clang compiled locally that changes commit-to-commit.

cachemeifyoucan commented on clang/lib/Driver/ToolChains/Darwin.cpp:
```diff
@@ -759,9 +855,27 @@ void darwin::Linker::ConstructJob(Compilation &C, const JobAction &JA,
 
+  // Set the environment to include ZERO_AR_FLAGS.
+  //
+  // FIXME: Add a flag to the linkers to support this without an environment
```
Another reproducibility related upstreaming. Do we need this?

akyrtzi replied:
Seems useful.

cachemeifyoucan commented on clang/tools/libclang/CDependencies.cpp:
```diff
@@ -53,8 +53,11 @@ void clang_experimental_ModuleDependencySet_dispose(
+  IntrusiveRefCntPtr<llvm::cas::CachingOnDiskFileSystem> FS;
   return wrap(new DependencyScanningService(
-      ScanningMode::DependencyDirectivesScan, unwrap(Format),
+      ScanningMode::DependencyDirectivesScan, unwrap(Format), CASOpts, FS,
```
The fix in this file is just get things compiling. I don't think we need them anymore

cachemeifyoucan reviewed (commented):
@swift-ci please test

cachemeifyoucan commented:
@swift-ci Please smoke test

cachemeifyoucan committed:
Add LLVM_ENABLE_ONDISK_CAS configuration

The option can be used to disable OnDiskCAS on windows because the
implementation is not compatible with windows yet.

benlangmuir commented on clang/include/clang/Frontend/CompilerInvocation.h:
```diff
@@ -242,6 +252,56 @@ class CompilerInvocation : public CompilerInvocationRefBase,
+  CASOptions &getCASOpts() { return CASOpts; }
+  const CASOptions &getCASOpts() const { return CASOpts; }
+
+  AnalyzerOptionsRef getAnalyzerOpts() const { return AnalyzerOpts; }
```
Do you know where these extra accessors came from? We should already inherit all of them (except CASOpts) from `CompilerInvocationRefBase` or `CompilerInvocationValueBase`.

benlangmuir commented on llvm/include/llvm/CAS/CASDB.h:
```diff
@@ -0,0 +1,528 @@
+using BlobProxy = LeafNodeProxy;
+
+Expected<std::unique_ptr<CASDB>>
+createPluginCAS(StringRef PluginPath, ArrayRef<std::string> PluginArgs = None);
```
This unused function and the (dummy) implementation in lib/CAS/PluginCAS.cpp should be removed.

benlangmuir commented on llvm/include/llvm/Support/raw_ostream_proxy.h:
```diff
@@ -0,0 +1,158 @@
+
+/// Non-owning proxy for a \a raw_ostream. Enables passing a stream into of an
+/// API that takes ownership.
+class raw_ostream_proxy : public raw_ostream_proxy_adaptor<> {
```
Is `raw_ostream_proxy` used anywhere? I only see `raw_pwrite_stream_proxy` being used. Maybe we can drop this one?

benlangmuir commented on llvm/include/llvm/Support/Path.h:
```diff
@@ -48,6 +48,14 @@ constexpr bool is_style_posix(Style S) {
 /// Check if \p S uses Windows path rules.
 constexpr bool is_style_windows(Style S) { return !is_style_posix(S); }
 
+constexpr Style system_style() {
```
The only place this is used is to check if it's posix.  Instead we should check `is_style_posix(Style::native)` which already does the right thing, and lets us avoid modifying Path.h.

benlangmuir commented on clang/tools/driver/clang-cache-build-session:
```diff
@@ -0,0 +1,9 @@
+# Uses the process id to ensure parallel invocations of this script will not
+# share the same identifier, and 'elapsed seconds since epoch' to ensure the
+# same for consecutive invocations.
+env CLANG_CACHE_BUILD_SESSION_ID=$$-$(date +%s) "$@"
```
This is the old version before it became a real binary.  Maybe need to update the PR?

cachemeifyoucan committed:
Update CachingOnDiskFileSystem to built on Windows

benlangmuir commented on clang/tools/driver/driver.cpp:
```diff
@@ -50,6 +52,18 @@
+#define USE_APPLE_LIBPROC_FOR_DEPSCAN_ANCESTORS
+#include <libproc.h>
+#include <sys/proc_info.h>
+#endif
```
Are all these headers being used somewhere? I think this change may be dead.

benlangmuir commented on clang/tools/driver/CMakeLists.txt:
```diff
@@ -82,6 +111,7 @@ if (APPLE)
   set(TOOL_INFO_PLIST_OUT "${CMAKE_CURRENT_BINARY_DIR}/${TOOL_INFO_PLIST}")
   target_link_libraries(clang
     PRIVATE
+    -lutil # FIXME: Apple-only. Not portable
```
What are we using from this library?

cachemeifyoucan committed:
MSVC compatibility fix

cachemeifyoucan committed:
Address review feedback

Cleanup some unneccessary diff from the branch.

cachemeifyoucan commented:
Update

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan committed:
[CMake] LLVM_ENABLE_ONDISK_CAS default fix on Windows

Fix the default of the LLVM_ENABLE_ONDISK_CAS is not correctly reset on
Windows to disable OnDiskCAS that is not yet compatible with windows.

cachemeifyoucan committed:
Get cache-build-session building on windows

cachemeifyoucan committed:
Disable cc1depscan on Windows

cc1depscan relies on posix and socket currently. Disable on Windows.

cachemeifyoucan commented:
@swift-ci please test
